clevertap-core/src/main/java/com/clevertap/android/sdk/LocalDataStore.java

```diff
@@ -115,6 +135,213 @@ public void persistEvent(Context context, JSONObject event, int type) {
+        DBAdapter dbAdapter = baseDatabaseManager.loadDBAdapter(context);
+        boolean updatedEventByDeviceID = dbAdapter.userEventLogDAO().updateEventByDeviceID(deviceID, eventName);
+        getConfigLogger().verbose("updatedEventByDeviceID = "+updatedEventByDeviceID);
+        return updatedEventByDeviceID;
+    }
+
+    @WorkerThread
+    public boolean updateUserEventLog(String eventName) {
+        String deviceID = deviceInfo.getDeviceID();
+        return updateEventByDeviceID(deviceID,eventName);
+    }
+
+    @WorkerThread
+    public boolean upsertUserEventLogsInBulk(Set<String> eventNames){
+        String deviceID = deviceInfo.getDeviceID();
+        DBAdapter dbAdapter = baseDatabaseManager.loadDBAdapter(context);
+        boolean upsertEventByDeviceID = dbAdapter.userEventLogDAO().upsertEventsByDeviceID(deviceID, eventNames);
+        getConfigLogger().verbose("upsertEventByDeviceID = "+upsertEventByDeviceID);
+        return upsertEventByDeviceID;
+    }
+
+    @WorkerThread
+    private long insertEventByDeviceID(String deviceID, String eventName) {
+        DBAdapter dbAdapter = baseDatabaseManager.loadDBAdapter(context);
+        long rowId = dbAdapter.userEventLogDAO().insertEventByDeviceID(deviceID, eventName);
+        getConfigLogger().verbose("inserted rowId = "+rowId);
+        return rowId;
+    }
+
+    @WorkerThread
+    public boolean insertUserEventLog(String eventName) {
+        String deviceID = deviceInfo.getDeviceID();
+        long rowId = insertEventByDeviceID(deviceID, eventName);
+        return rowId >= 0;
+    }
+
+    @WorkerThread
+    private boolean eventExistsByDeviceID(String deviceID, String eventName) {
+        DBAdapter dbAdapter = baseDatabaseManager.loadDBAdapter(context);
+        boolean eventExistsByDeviceID = dbAdapter.userEventLogDAO().eventExistsByDeviceID(deviceID, eventName);
+        getConfigLogger().verbose("eventExistsByDeviceID = "+eventExistsByDeviceID);
+        return eventExistsByDeviceID;
+    }
+
+    @WorkerThread
+    public boolean isUserEventLogExists(String eventName) {
+        String deviceID = deviceInfo.getDeviceID();
+        return eventExistsByDeviceID(deviceID,eventName);
+    }
+
+    @WorkerThread
+    private boolean eventExistsByDeviceIDAndCount(String deviceID, String eventName, int count) {
+        DBAdapter dbAdapter = baseDatabaseManager.loadDBAdapter(context);
+        boolean eventExistsByDeviceIDAndCount = dbAdapter.userEventLogDAO()
+                .eventExistsByDeviceIDAndCount(deviceID, eventName, count);
+
+        getConfigLogger().verbose("eventExistsByDeviceIDAndCount = "+eventExistsByDeviceIDAndCount);
+        return eventExistsByDeviceIDAndCount;
+    }
+
+    @WorkerThread
+    public boolean isUserEventLogFirstTime(String eventName) {
+        if (userEventLogKeys.contains(eventName)) {
+            return false;
+        }
+
+        String deviceID = deviceInfo.getDeviceID();
```

Comment:
can we cache device id so retrieval does not happen frequently from shared prefs?

Reply:
I checked code and seems not feasible currently. However, there is a long term solution that can be implemented as a separate task. The idea is to cache device id in `DeviceCachedInfo` and on user change, update value in `DeviceCachedInfo`, but this change will reflect among all feature/entire codebase which will require extensive testing.